sdk/src/main/scala/io/horizen/account/websocket/WebSocketAccountServerEndpoint.scala

```diff
@@ -0,0 +1,270 @@
+
+      case unkownMethod =>
+        WebSocketAccountServerEndpoint.send(new RpcResponseError(rpcRequest.id,
+          new RpcError(RpcCode.InvalidParams, "unsupported subscription type " + unkownMethod, "")),
+          session)
+        log.debug("unsupported subscription type " + unkownMethod)
+    }
+
+  }
+
+  private def unsubscribe(session: Session, rpcRequest: RpcRequest): Unit = {
+    val rpcParams = rpcRequest.params
+
+    val subscriptionIdsToRemove: util.ArrayList[BigInteger] = new util.ArrayList[BigInteger]()
+    rpcParams.forEach(subscriptionId => {
+      subscriptionIdsToRemove.add(Numeric.decodeQuantity(subscriptionId.asText()))
+    })
+    val removedSubscription = WebSocketAccountServerEndpoint.removeSubscriptions(subscriptionIdsToRemove.toArray(new Array[BigInteger](0)))
+    if (!removedSubscription) {
+      WebSocketAccountServerEndpoint.send(new RpcResponseError(rpcRequest.id,
+        new RpcError(RpcCode.InvalidParams, s"Subscription ID not found.", "")),
+        session)
+    } else
+      WebSocketAccountServerEndpoint.send(new RpcResponseSuccess(rpcRequest.id, true), session)
+  }
+
+  private def validateRpcRequestParams(rpcParams: JsonNode, rpcRequest: RpcId, session: Session): Unit = {
+    if (rpcParams == null || !rpcParams.isArray || rpcParams.size() < 1) {
+      log.debug("Missing or empty field params.")
+      WebSocketAccountServerEndpoint.send(new RpcResponseError(rpcRequest,
+        new RpcError(RpcCode.InvalidParams, "Missing or empty field params.", "")),
+        session)
+    }
+  }
+
+  private def createSubscriptionId(): BigInteger = {
+    BigInteger.valueOf(WebSocketAccountServerEndpoint.subscriptionCounter.incrementAndGet())
+  }
+}
+
+private object WebSocketAccountServerEndpoint extends SparkzLogging {
+  var subscriptionCounter: AtomicInteger = new AtomicInteger(0)
+  var newHeadsSubscriptions: util.ArrayList[Subscription] = new util.ArrayList[Subscription]()
+  var newPendingTransactionsSubscriptions: util.ArrayList[Subscription] = new util.ArrayList[Subscription]()
+  var logsSubscriptions: util.ArrayList[SubscriptionWithFilter] = new util.ArrayList[SubscriptionWithFilter]()
+
+  val webSocketAccountChannelImpl = new WebSocketAccountChannelImpl()
+  private var walletAddresses: Set[Address] = webSocketAccountChannelImpl.getWalletAddresses
+  private var cachedBlocksReceipts: mutable.Queue[(String, Set[EthereumLogView])] = new mutable.Queue[(String, Set[EthereumLogView])]()
+  private val maxCachedBlockReceipts = 100
+
+  def notifySemanticallySuccessfulModifier(block: AccountBlock): Unit = {
+    log.info("Websocket received new block: "+block.toString)
+
+    val blockJson = webSocketAccountChannelImpl.accountBlockToWebsocketJson(block)
+
+    newHeadsSubscriptions.forEach(subscription => {
+      send(new WebSocketAccountEvent(params = new WebSocketAccountEventParams(subscription.subscriptionId, blockJson)), subscription.session)
+    })
+
+    //We have a chain reorganization
```

Comment:
let's put this comment inside `while` loop, because here we just check IF we have reorg

Reply:
Done